feat: deserialize form data objects

alcidesqueiroz opened the pull request:
closes #429 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

mcollina reviewed (approved):
lgtm

climba03003 reviewed (changes requested):
Thanks for the PR.

Can you have any reference on the implementation syntax for `array` and `json` key structure?
For example, I never see `foo{}` key as a indicator for `JSON` value.

How would you distinguish if `foo[0]` means to be `array` but not `object` property?

And final spaces is y would need to write the document for new feature before it get merged.

alcidesqueiroz commented:
Hi, @climba03003. 

I've just updated the PR with the requested changes. In order to keep backward compatibility, I disabled key special notation parsing by default. To enable it, I introduced an option called `enableSpecialNotationKeys` that might be set either to `true` or to `objectsOnly` (similarly to the existing `attachFieldsToBody` option, that can be either `true` or `keyValues`).

Answering your previous questions:

1. I borrowed this particular notation from [Axios serializer](https://axios-http.com/docs/multipart#automatic-serialization)
2. My first approach was to parse numeric keys as array indexes. However, after seeing how `qs` deals with this dilemma, I've adopted a similar strategy, allowing the developer to disable this behavior through an option. For this, the `enableSpecialNotationKeys` option can be set to `objectsOnly`.
3. Good point. I changed my implementation to prevent it.
4. I've just included complete documentation for this feature.

simoneb commented on index.js:
```diff
@@ -83,6 +93,106 @@ function attachToBody (options, req, reply, next) {
   })
 }
 
+function validateSerializedKey (tokens, value) {
```
hard to follow what's going on here but I guess this is something that we have to do. I have a feeling that this code may be coming from somewhere else, possibly slightly modified though. if that is the case, don't forget to cite the source please

alcidesqueiroz replied:
Even though I originally wanted to save time by starting from some pre-existing battle-tested code, I couldn't find anything close to what I needed, so I ended up writing the entire feature from scratch. Zero third-party code. 😄

climba03003 replied:
Then it doesn't make sense why doesn't use a single `for-loop`.

simoneb commented on README.md:
````diff
@@ -517,6 +517,125 @@ fastify.post('/upload/files', {
 })
 ```
 
+## Parsing keys with nested arrays and objects
````
great docs 👌

alcidesqueiroz replied:
🙇🏻‍♂️

climba03003 commented on index.js:
```diff
@@ -83,6 +93,106 @@ function attachToBody (options, req, reply, next) {
+    if (token === ']' && prevToken === '[') {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? (parseArrays ? [] : {})
+      key = parseArrays ? parentRef.length : Object.keys(parentRef).length
+    } else if (parseArrays && token === ']' && /^\d+$/.test(prevToken)) {
```
prevToken is a single character.
It is necessary to use regexp for this check?

alcidesqueiroz replied:
In this case, `prevToken` not necessarily will have a single character. This regex would also match a multi-digit number. But thanks to your comment, I remembered that I was not checking for leading zeroes, so I extracted this checking to a function and explicitly checked if the number has no leading zero. 

I've also [benchmarked this regex-based checking against two other alternatives](https://perf.link/#eyJpZCI6IndjcXoyZDM5OTFwIiwidGl0bGUiOiJDaGVja2luZyBpZiBzdHJpbmcgaXMgYSB2YWxpZCBpbnRlZ2VyIG51bWJlciA%2BPSAwIiwiYmVmb3JlIjoiY29uc3QgZGF0YSA9IFsnMCcsICcwMDEnLCAnMTAwJywgJzEyMzQnLCAnMjInLCAnOTknLCAnYWJjJywgJy0xMjMnLCAnMS43MicsICcxejMnXTtcblxuZnVuY3Rpb24gaXNWYWxpZEluZGV4VjEoc3RyKSB7XG4gIGlmIChzdHIubGVuZ3RoID4gMSAmJiBzdHJbMF0gPT09ICcwJykgcmV0dXJuIGZhbHNlO1xuICByZXR1cm4gL15cXGQrJC8udGVzdChzdHIpO1xufVxuXG5mdW5jdGlvbiBpc1ZhbGlkSW5kZXhWMihzdHIpIHtcbiAgaWYgKHN0ci5sZW5ndGggPiAxICYmIHN0clswXSA9PT0gJzAnKSByZXR1cm4gZmFsc2U7XG4gIGZvciAobGV0IGkgPSAwOyBpIDwgc3RyLmxlbmd0aDsgaSsrKSB7XG4gICAgY29uc3QgYyA9IHN0ci5jaGFyQ29kZUF0KGkpO1xuICAgIGlmIChjIDwgNDggfHwgYyA%2BIDU3KSByZXR1cm4gZmFsc2U7XG4gIH1cbiAgcmV0dXJuIHRydWU7XG59XG5cbmZ1bmN0aW9uIGlzVmFsaWRJbmRleFYzKHN0cikge1xuICBpZiAoc3RyWzBdID09PSAnLScpIHJldHVybiBmYWxzZTtcbiAgcmV0dXJuIHBhcnNlRmxvYXQoc3RyKS50b1N0cmluZygpID09PSBzdHIudG9TdHJpbmcoKTtcbn0iLCJ0ZXN0cyI6W3sibmFtZSI6IlJlZ2V4LWJhc2VkIiwiY29kZSI6ImRhdGEuZm9yRWFjaChzdHIgPT4gaXNWYWxpZEluZGV4VjEoc3RyKSkiLCJydW5zIjpbMjg2MDAwLDEzMDAwMCwxMDAwLDcyMDAwLDE1MTAwMCwxMDAwLDIyNjAwMCwzODAwMCwxODMwMDAsMjAxMDAwLDMyMTAwMCwyMjYwMDAsMTc4MDAwLDg0MDAwLDI5NTAwMCwxMDIwMDAsMzU1MDAwLDIxMDAwMCwzNzAwMCw5MzAwMCw4NzAwMCwyMzQwMDAsMjQ2MDAwLDIwMDAsMTAwMCwxMjMwMDAsMTk4MDAwLDE5ODAwMCwxNDAwMCwxMDAwLDEwMDAsMjAwMCwxOTcwMDAsNDYwMDAsMTAwMCwyMzMwMDAsMjI4MDAwLDMyMjAwMCwxODMwMDAsMTAxMDAwLDMyNjAwMCwyNDcwMDAsMjE0MDAwLDE4MTAwMCw2NTAwMCwxMDAwLDM1NTAwMCwxNjAwMCwxMDAwLDEwOTAwMCwyMTAwMCwzNTUwMDAsMTAwMCwyODAwMDAsMjA3MDAwLDIyMjAwMCwxMzkwMDAsMjIxMDAwLDMyMDAwLDExNDAwMCwyNTkwMDAsMTkwMDAwLDIzOTAwMCwxMTMwMDAsMTAwMCwyODUwMDAsMzQxMDAwLDI0NjAwMCw4MzAwMCwzMTIwMDAsMTUwMDAwLDE5MTAwMCwyMTcwMDAsMTkxMDAwLDEwMDAsOTMwMDAsMjQwMDAsMzE5MDAwLDIxNDAwMCwzNTUwMDAsMzU1MDAwLDQyMDAwLDEwMDAsMzcyMDAwLDEwMDAsNjgwMDAsMjg5MDAwLDEwMDAsNzEwMDAsMzEwMDAwLDIxOTAwMCw3OTAwMCwzNTUwMDAsMTAwMCwxMDAwLDI5NzAwMCwxMDAwLDI3OTAwMCwyNjUwMDAsMTc5MDAwXSwib3BzIjoxNTcyNTB9LHsibmFtZSI6Ikl0ZXJhdGlvbi1iYXNlZCIsImNvZGUiOiJkYXRhLmZvckVhY2goc3RyID0%2BIGlzVmFsaWRJbmRleFYyKHN0cikpIiwicnVucyI6WzI0MjAwMCwzNjAwMCwzMDAwLDE0NzAwMCwyOTAwMCwyNzMwMDAsMTUzMDAwLDYwMDAwLDgyMDAwLDI3NjAwMCwxMzQwMDAsMTQ1MDAwLDIyMDAwLDE0MTAwMCwzMTYwMDAsNjYwMDAsMjA4MDAwLDEyNzAwMCwxMzIwMDAsMTU4MDAwLDEzMjAwMCw5MjAwMCw0NDAwMCwxMDAwLDIzNjAwMCwxMjIwMDAsMTc3MDAwLDMyMzAwMCwxMzIwMDAsMzE4MDAwLDI3MjAwMCwzMTgwMDAsMTU4MDAwLDQ1MDAwLDEwMDAsMTAwMCwyNDUwMDAsODMwMDAsMTAwMCwxMTEwMDAsMjM5MDAwLDEwMDAsMTEzMDAwLDIwMDAwLDIxNDAwMCwyMzQwMDAsMjEzMDAwLDEzMjAwMCwxMzAwMCw0ODAwMCwyMjUwMDAsMTMyMDAwLDIxNjAwMCwxOTcwMDAsMTI3MDAwLDE2NjAwMCw2NDAwMCwxMzIwMDAsMjI4MDAwLDExMTAwMCwxMDcwMDAsMjIyMDAwLDk5MDAwLDIwMDAsMjE1MDAwLDIyNjAwMCwyNzAwMDAsMTU0MDAwLDI4MTAwMCwxMzIwMDAsNDQwMDAsMTY0MDAwLDEwMzAwMCwzMjAwMDAsMjUxMDAwLDkwMDAwLDIwNDAwMCwzMjAwMCw3MjAwMCwxOTkwMDAsMjA2MDAwLDE1NzAwMCwxMzgwMDAsMjAwMDAsMjQ3MDAwLDEzMjAwMCwzODAwMCwxMzUwMDAsMTQ2MDAwLDM1MDAwLDEzMjAwMCw1MzAwMCwxOTAwMCwxMDAwLDEzMjAwMCwxMjYwMDAsMjQxMDAwLDIyNjAwMCwxNzcwMDAsMTMwMDAwXSwib3BzIjoxNDA2NTB9LHsibmFtZSI6InBhcnNlRmxvYXQtYmFzZWQiLCJjb2RlIjoiZGF0YS5mb3JFYWNoKHN0ciA9PiBpc1ZhbGlkSW5kZXhWMyhzdHIpKSIsInJ1bnMiOlsyMzYwMDAsMTAwMCwxMDAwLDMxODAwMCwxMjQwMDAsMTgxMDAwLDEwMDAsMjM0MDAwLDI3NzAwMCwyNjUwMDAsMTk4MDAwLDY5MDAwLDcwMDAsMzU2MDAwLDE4MzAwMCwzNDAwMCwzNjIwMDAsMjAwMCw2MDAwLDI0NTAwMCwzNTAwMCwyNDYwMDAsMzMwMDAsMzAzMDAwLDIzNjAwMCwyOTIwMDAsMjExMDAwLDE3MDAwMCwxMDAwLDcwMDAsMjU0MDAwLDI2NTAwMCwzMzEwMDAsMjQ1MDAwLDEwMDAsMjgxMDAwLDE4NDAwMCwxOTYwMDAsMTAwMCw4NDAwMCwxMDAwLDIxOTAwMCwxNjYwMDAsMTAwMCwxMTIwMDAsNDYwMDAsMTIxMDAwLDE3MzAwMCwxODIwMDAsMjI3MDAwLDY1MDAwLDIyMjAwMCwxMDAwLDE3NDAwMCwzMjAwMDAsMjU1MDAwLDE3NDAwMCwyNzcwMDAsMjE1MDAwLDEwMDAsMTAwMCwxODcwMDAsMjczMDAwLDEwMDAsMjQ5MDAwLDE2MjAwMCwyNjIwMDAsOTAwMCwxMDAwLDEyMjAwMCwxMDAwLDEwMDAsMTc0MDAwLDIzMDAwLDEwMDAsMzA4MDAwLDI5MDAwLDEwMDAsMTYwMDAwLDI0ODAwMCwyOTQwMDAsMTAwMCwxNDUwMDAsNjgwMDAsODYwMDAsNTEwMDAsMjkwMDAsMjYzMDAwLDIzMDAwMCwxNzMwMDAsMjI2MDAwLDMzMTAwMCwyMDAwMDAsMTAxMDAwLDc4MDAwLDEwMDAsMjc5MDAwLDE0NTAwMCwxNzQwMDAsMjMwMDAwXSwib3BzIjoxNDc0NjB9XSwidXBkYXRlZCI6IjIwMjMtMDUtMDlUMTg6NDY6MTIuNTkyWiJ9), one iteration-based and other that relies on parseFloat. The regex version won by a slim margin.

climba03003 replied:
Unless I read the code wrongly `const prevToken = tokens[i - 1]`.
`prevToken` is never reassigned and is always a single character.
That means matching multi-digit number is useless for `prevToken`.

alcidesqueiroz replied:
`tokens` is not a string, it's an array of strings. So, `prevToken` might be either `[`, `]`, `{}`, a multi-digit numeric index or an alphanumeric object key. Besides, `prevToken` gets reassigned on each iteration.

climba03003 replied:
Then it doesn't make sense why doesn't use a single for-loop.

alcidesqueiroz replied:
The logic I've proposed relies on more than a loop in order to separate concerns. Mixing all steps in a single loop would make it way harder to read, debug and maintain.

alcidesqueiroz commented on index.js:
```diff
@@ -73,7 +73,17 @@ function attachToBody (options, req, reply, next) {
     }
-    if (body[key] === undefined) {
+
+    if ([true, 'objectsOnly'].includes(options.enableSpecialNotationKeys) && /(\[|]|{})/.test(key)) {
```
I implemented an `O(n)` tokenization logic and [benchmarked it](https://perf.link/#eyJpZCI6InNzYnlldHFlbGx5IiwidGl0bGUiOiJQYXJzaW5nIGEga2V5IGFuZCBleHRyYWN0aW5nIGl0cyB0b2tlbnMiLCJiZWZvcmUiOiJjb25zdCBkYXRhID0gWydwcm9wQVtwcm9wQUFdJywgJ3Byb3BBW3Byb3BBQl0nLCAncHJvcEFbcHJvcEFDXVtdJywgJ3Byb3BBW3Byb3BBQ11bXScsICdwcm9wQVtwcm9wQUNdW10nLCAncHJvcEJbMF1bcHJvcEJBXVtwcm9wQkFBXScsICdwcm9wQlswXVtwcm9wQkFdW3Byb3BCQUJdJywgJ3Byb3BCWzBdW3Byb3BCQV1bcHJvcEJBQ11bMF0nLCAncHJvcEJbMF1bcHJvcEJBXVtwcm9wQkFDXVsxXScsICdwcm9wQlswXVtwcm9wQkFdW3Byb3BCQUNdWzJdJywgJ3Byb3BCWzBdW3Byb3BCQl0nLCAncHJvcEJbMV1bcHJvcEJBXVtwcm9wQkFBXScsICdwcm9wQ1tdJywgJ3Byb3BDW10nLCAncHJvcENbXScsICdwcm9wRCcsICdwcm9wRScsICdwcm9wRScsICdwcm9wRScsICdwcm9wRnt9JywgJ3Byb3BHWzBdW3Byb3BHQV1bXXt9JywgJ3Byb3BHWzFde30nXTtcblxuZnVuY3Rpb24gZ2V0U2VyaWFsaXplZEtleVRva2Vuc1YxKGtleSkge1xuICBjb25zdCB0b2tlbnMgPSBrZXkuc3BsaXQoLyhcXFt8XXx7fSkvKS5maWx0ZXIoQm9vbGVhbik7XG4gIHJldHVybiB0b2tlbnMuc29tZSh0b2tlbiA9PiBbJ1snLCAnXScsICd7fSddLmluY2x1ZGVzKHRva2VuKSkgPyB0b2tlbnMgOiBbXTtcbn1cblxuZnVuY3Rpb24gZ2V0U2VyaWFsaXplZEtleVRva2Vuc1YyKGtleSkge1xuICBsZXQgaGFzU3BlY2lhbE5vdGF0aW9uID0gZmFsc2U7XG4gIGxldCBzdGFydEZyb20gPSAwO1xuICBsZXQgaSA9IDA7XG4gIGNvbnN0IHRva2VucyA9IFtdO1xuXG4gIHdoaWxlIChpIDwga2V5Lmxlbmd0aCkge1xuICAgIGxldCB0b2tlbkxlbmd0aDtcbiAgICBjb25zdCBjaGFyID0ga2V5W2ldO1xuICAgIFxuICAgIGlmIChjaGFyID09PSAnWycgfHwgY2hhciA9PT0gJ10nKSB7XG4gICAgICB0b2tlbkxlbmd0aCA9IDE7XG4gICAgfSBlbHNlIGlmIChjaGFyID09PSAneycgJiYga2V5W2kgKyAxXSA9PT0gJ30nKSB7XG4gICAgICB0b2tlbkxlbmd0aCA9IDI7XG4gICAgfSBcblxuICAgIGlmICh0b2tlbkxlbmd0aCkge1xuICAgICAgaGFzU3BlY2lhbE5vdGF0aW9uID0gdHJ1ZTtcbiAgICAgIGNvbnN0IHByZXZUb2tlbiA9IGtleS5zbGljZShzdGFydEZyb20sIGkpO1xuICAgICAgaWYgKHByZXZUb2tlbikgdG9rZW5zLnB1c2gocHJldlRva2VuKTtcbiAgICAgIHRva2Vucy5wdXNoKGtleS5zbGljZShpLCBpICsgdG9rZW5MZW5ndGgpKTtcbiAgICAgIHN0YXJ0RnJvbSA9IGkgKyB0b2tlbkxlbmd0aDtcbiAgICAgIGkgPSBzdGFydEZyb207XG4gICAgICBjb250aW51ZTtcbiAgICB9XG5cbiAgICBpZiAoaSA9PT0ga2V5Lmxlbmd0aCAtIDEpIHtcbiAgICAgIGNvbnN0IHByZXZUb2tlbiA9IGtleS5zbGljZShzdGFydEZyb20pO1xuICAgICAgaWYgKHByZXZUb2tlbikgdG9rZW5zLnB1c2gocHJldlRva2VuKTtcbiAgICB9XG5cbiAgICBpKys7XG4gIH1cblxuICByZXR1cm4gaGFzU3BlY2lhbE5vdGF0aW9uID8gdG9rZW5zIDogW107XG59IiwidGVzdHMiOlt7Im5hbWUiOiJSZWdleC1iYXNlZCIsImNvZGUiOiJkYXRhLmZvckVhY2goa2V5ID0%2BIGdldFNlcmlhbGl6ZWRLZXlUb2tlbnNWMShrZXkpKSIsInJ1bnMiOlsxNTAwMCwxMDAwLDI1MDAwLDEwMDAsMTAwMCwyNDAwMCwxODAwMCwxNjAwMCwyNTAwMCw5MDAwLDM3MDAwLDcwMDAsMTAwMCwyNDAwMCwyNDAwMCwxMDAwLDEyMDAwLDkwMDAsMTgwMDAsMTAwMCwxMDAwLDI2MDAwLDIyMDAwLDIyMDAwLDI0MDAwLDI5MDAwLDQwMDAsMjEwMDAsMjUwMDAsMzMwMDAsMTAwMCwyNTAwMCwxMTAwMCwxMDAwLDUwMDAsMjAwMCwyMTAwMCw4MDAwLDEwMDAsNDAwMCwxMDAwLDIyMDAwLDEwMDAsMTAwMCwyMDAwMCwyNDAwMCwxMDAwLDQwMDAsMTAwMCwxMDAwLDE1MDAwLDE3MDAwLDExMDAwLDE3MDAwLDMyMDAwLDI3MDAwLDQwMDAsMzAwMCw3MDAwLDM1MDAwLDMxMDAwLDM1MDAwLDgwMDAsMjIwMDAsMTAwMCwyMzAwMCwxOTAwMCwxMDAwLDIyMDAwLDEwMDAsMjMwMDAsMTAwMCwyMDAwMCwzMDAwLDI4MDAwLDEwMDAsNjAwMCwyMTAwMCwxODAwMCwxMDAwLDE2MDAwLDMxMDAwLDM1MDAwLDMwMDAsMTAwMCwzNDAwMCwzNDAwMCwxMDAwLDE3MDAwLDIwMDAwLDE2MDAwLDEwMDAsMzQwMDAsMTMwMDAsMjYwMDAsMTMwMDAsODAwMCwxMDAwLDE0MDAwLDMxMDAwXSwib3BzIjoxNDMzMH0seyJuYW1lIjoiSXRlcmF0aW9uLWJhc2VkIiwiY29kZSI6ImRhdGEuZm9yRWFjaChrZXkgPT4gZ2V0U2VyaWFsaXplZEtleVRva2Vuc1YyKGtleSkpIiwicnVucyI6WzEyMDAwLDEwMDAsMTIwMDAsNTAwMCw2MDAwLDEyMDAwLDEyMDAwLDIwMDAsNDAwMCw1MDAwLDEyMDAwLDQwMDAsMTMwMDAsNzAwMCwxMzAwMCwyMDAwLDEyMDAwLDEyMDAwLDEwMDAsOTAwMCwxNTAwMCwxMDAwLDEwMDAsNDAwMCwxMjAwMCw1MDAwLDEyMDAwLDYwMDAsMjAwMCwxMjAwMCwxMDAwLDYwMDAsNjAwMCwxMTAwMCwyMDAwLDUwMDAsMTAwMCwxODAwMCwxMjAwMCw4MDAwLDQwMDAsMTAwMCw1MDAwLDEwMDAwLDkwMDAsODAwMCwxMzAwMCwxNTAwMCwxMjAwMCwxNTAwMCwxMDAwLDIwMDAsMTYwMDAsMTYwMDAsNDAwMCw1MDAwLDUwMDAsNjAwMCw5MDAwLDE0MDAwLDcwMDAsNTAwMCw0MDAwLDE4MDAwLDE0MDAwLDkwMDAsNjAwMCwxNDAwMCw2MDAwLDEyMDAwLDcwMDAsNDAwMCwxNjAwMCw5MDAwLDQwMDAsMTcwMDAsMTcwMDAsMTIwMDAsMTAwMCw4MDAwLDEyMDAwLDUwMDAsMTEwMDAsMTIwMDAsMzAwMCwxMDAwLDcwMDAsMTAwMDAsMTIwMDAsMTAwMCwyMDAwLDIwMDAsMTUwMDAsMTAwMCw3MDAwLDEwMDAsMTYwMDAsNjAwMCw0MDAwLDYwMDBdLCJvcHMiOjc5MDB9XSwidXBkYXRlZCI6IjIwMjMtMDUtMDlUMTg6NTQ6MjQuOTc3WiJ9) against the regex-based version. The regex implementation won. I've improved my original code, so the regex gets executed just once and only if the `enableSpecialNotationKeys` option is  on.

Uzlopak commented on index.js:
```diff
@@ -83,6 +94,116 @@ function attachToBody (options, req, reply, next) {
+}
+
+function setSerializedKey (segments) {
+  segments.reverse()
```
instead of reversing, it would make more sense, to change the for loop

maybe something like this. not tested. possibly wrong.
```js
function setSerializedKey (segments) {
  const len = segments.length

  for (let i = len; i >= 0; --i) {
    const segment = segments[i]
    const prevSegment = segments[i - 1]
    value = prevSegment ? prevSegment.parentRef : value
    segment.parentRef[segment.key] = value
  }
}
```

Uzlopak commented on index.js:
```diff
@@ -83,6 +94,116 @@ function attachToBody (options, req, reply, next) {
+}
+
+function isValidIndex (str) {
+  if (str.length > 1 && str[0] === '0') return false
```
Maybe a for loop is faster than the regex.

Not tested, maybe wrong. Can  be used for inspiration
```js
const numbersAsciiLookup = Array(256).fill(0).map((v,i) => /^\d+$/.test(String.fromCharCode(i)) && true || false)

function isValidIndex (idx) {
  const len = idx.length
  let result = true
  let i = 0
  let code = 0
  while (i < len) {
    code = idx.charCodeAt(i)
    if (!numbersLookup[code]) {
     result = false
     break;
    }
    if (i === 0 && code === 48 /* "0" */) {
      result = false
      break;
    }
    ++i
  }  
  return result
}
```

alcidesqueiroz replied:
[I had already benchmarked](https://perf.link/#eyJpZCI6IndjcXoyZDM5OTFwIiwidGl0bGUiOiJDaGVja2luZyBpZiBzdHJpbmcgaXMgYSB2YWxpZCBpbnRlZ2VyIG51bWJlciA%2BPSAwIiwiYmVmb3JlIjoiY29uc3QgZGF0YSA9IFsnMCcsICcwMDEnLCAnMTAwJywgJzEyMzQnLCAnMjInLCAnOTknLCAnYWJjJywgJy0xMjMnLCAnMS43MicsICcxejMnXTtcblxuZnVuY3Rpb24gaXNWYWxpZEluZGV4VjEoc3RyKSB7XG4gIGlmIChzdHIubGVuZ3RoID4gMSAmJiBzdHJbMF0gPT09ICcwJykgcmV0dXJuIGZhbHNlO1xuICByZXR1cm4gL15cXGQrJC8udGVzdChzdHIpO1xufVxuXG5mdW5jdGlvbiBpc1ZhbGlkSW5kZXhWMihzdHIpIHtcbiAgaWYgKHN0ci5sZW5ndGggPiAxICYmIHN0clswXSA9PT0gJzAnKSByZXR1cm4gZmFsc2U7XG4gIGZvciAobGV0IGkgPSAwOyBpIDwgc3RyLmxlbmd0aDsgaSsrKSB7XG4gICAgY29uc3QgYyA9IHN0ci5jaGFyQ29kZUF0KGkpO1xuICAgIGlmIChjIDwgNDggfHwgYyA%2BIDU3KSByZXR1cm4gZmFsc2U7XG4gIH1cbiAgcmV0dXJuIHRydWU7XG59XG5cbmZ1bmN0aW9uIGlzVmFsaWRJbmRleFYzKHN0cikge1xuICBpZiAoc3RyWzBdID09PSAnLScpIHJldHVybiBmYWxzZTtcbiAgcmV0dXJuIHBhcnNlRmxvYXQoc3RyKS50b1N0cmluZygpID09PSBzdHIudG9TdHJpbmcoKTtcbn1cblxuY29uc3QgbnVtYmVyc0FzY2lpTG9va3VwID0gQXJyYXkoMjU2KVxuICAuZmlsbCgwKVxuICAubWFwKCh2LCBpKSA9PiAoL15cXGQrJC8udGVzdChTdHJpbmcuZnJvbUNoYXJDb2RlKGkpKSAmJiB0cnVlKSB8fCBmYWxzZSk7XG5cbmZ1bmN0aW9uIGlzVmFsaWRJbmRleFY0KHN0cikge1xuICBjb25zdCBsZW4gPSBzdHIubGVuZ3RoO1xuICBsZXQgcmVzdWx0ID0gdHJ1ZTtcbiAgbGV0IGkgPSAwO1xuICBsZXQgY29kZSA9IDA7XG4gIHdoaWxlIChpIDwgbGVuKSB7XG4gICAgY29kZSA9IHN0ci5jaGFyQ29kZUF0KGkpO1xuICAgIGlmICghbnVtYmVyc0FzY2lpTG9va3VwW2NvZGVdKSB7XG4gICAgICByZXN1bHQgPSBmYWxzZTtcbiAgICAgIGJyZWFrO1xuICAgIH1cbiAgICBpZiAoaSA9PT0gMCAmJiBsZW4gPiAxICYmIGNvZGUgPT09IDQ4IC8qIFwiMFwiICovKSB7XG4gICAgICByZXN1bHQgPSBmYWxzZTtcbiAgICAgIGJyZWFrO1xuICAgIH1cbiAgICArK2k7XG4gIH1cbiAgcmV0dXJuIHJlc3VsdDtcbn1cblxuIiwidGVzdHMiOlt7Im5hbWUiOiJUZXN0IENhc2UiLCJjb2RlIjoiZGF0YS5mb3JFYWNoKHN0ciA9PiBpc1ZhbGlkSW5kZXhWMShzdHIpKSIsInJ1bnMiOlsxNzUwMDAsMTI0MDAwLDM1NTAwMCwyOTQwMDAsMzYxMDAwLDQyMzAwMCwyNzAwMCwyOTcwMDAsNDAwMCwzMTIwMDAsMTk0MDAwLDIwNjAwMCwzNTUwMDAsMjY0MDAwLDM0NjAwMCwzMzIwMDAsMjIxMDAwLDE1OTAwMCw2OTAwMCwzOTAwMDAsMzU1MDAwLDIwMDAsMzU1MDAwLDMzNDAwMCwxNzMwMDAsODQwMDAsMTM3MDAwLDI3NjAwMCwxMDgwMDAsMjEwMDAwLDIwODAwMCwzNTUwMDAsMTAwMCwxNzIwMDAsMzg5MDAwLDg1MDAwLDUwMDAwLDIxMTAwMCw0MzIwMDAsODcwMDAsNTYwMDAsMTMxMDAwLDExOTAwMCwzNTgwMDAsMTQyMDAwLDE1MDAwLDIwMjAwMCw0MDMwMDAsMTAxMDAwLDIyODAwMCwzNTUwMDAsNDUyMDAwLDM5MDAwLDE4MzAwMCwyNDkwMDAsNDQ4MDAwLDM1NTAwMCw1MTAwMCwyNDcwMDAsMTE5MDAwLDMwODAwMCwyNTgwMDAsOTQwMDAsMTU2MDAwLDExMDAwLDIxMTAwMCwzMTEwMDAsNDE1MDAwLDQxNjAwMCwzMTgwMDAsODAwMDAsMjI4MDAwLDIzMTAwMCwzNTUwMDAsMjIwMDAsMzU1MDAwLDEwMDAsMzU1MDAwLDkzMDAwLDMxODAwMCwzMTIwMDAsMzI3MDAwLDEwMDAsMjk4MDAwLDI0NDAwMCwzNTUwMDAsMjcwMDAsNzMwMDAsMjA5MDAwLDM1NTAwMCwxNDYwMDAsMzU1MDAwLDI0MzAwMCw5OTAwMCwyNTEwMDAsMTQ5MDAwLDE2NzAwMCwxNTcwMDAsMzU1MDAwLDgyMDAwXSwib3BzIjoyMTkyNjB9LHsibmFtZSI6IlJlZ2V4LWJhc2VkIiwiY29kZSI6ImRhdGEuZm9yRWFjaChzdHIgPT4gaXNWYWxpZEluZGV4VjIoc3RyKSkiLCJydW5zIjpbNDgwMDAsNjMwMDAsMjYwMDAwLDE5ODAwMCwyOTIwMDAsMjczMDAwLDMwMDAwLDIwMTAwMCwxMDAwLDIzMjAwMCwxMzIwMDAsMTM1MDAwLDI3NzAwMCwxNTQwMDAsMTk0MDAwLDIzOTAwMCwxNDcwMDAsMTI2MDAwLDUwMDAwLDEzMjAwMCwyNjMwMDAsMTAwMCwyMjIwMDAsMjYwMDAsMjUwMDAsMTU3MDAwLDEwNzAwMCwxMDAwLDE5ODAwMCwxNjIwMDAsMTIwMDAwLDI2MTAwMCwyMTAwMCwyOTAwMCw3NDAwMCwxMDMwMDAsMzEwMDAwLDEzMjAwMCwzMzEwMDAsMzQwMDAsMzQxMDAwLDEwMDAwLDIzNzAwMCwyODIwMDAsMTY5MDAwLDE0MjAwMCwyMDAwLDI3NDAwMCwzMjgwMDAsMjM0MDAwLDIzNjAwMCwzNTkwMDAsMTA1MDAwLDI5MzAwMCwxODIwMDAsMjg2MDAwLDE1MzAwMCwyMTAwMCwxMjgwMDAsMjE1MDAwLDIxNDAwMCwxMzIwMDAsNzMwMDAsMTAwMCw3MDAwLDExNTAwMCwxNzIwMDAsMzE1MDAwLDE2MTAwMCwyOTcwMDAsMTY0MDAwLDI4MjAwMCwyMTEwMDAsMTk0MDAwLDE1NTAwMCwzNTAwMCwxMDAwLDEzMjAwMCw4NjAwMCwyMjMwMDAsMjUzMDAwLDEzMjAwMCwyODgwMDAsMTYxMDAwLDE4MDAwMCwyMzMwMDAsMjIwMDAsMTAwMCwxMzIwMDAsMjc2MDAwLDExODAwMCwzMTAwMDAsMjE0MDAwLDI0ODAwMCwxMzIwMDAsMTA1MDAwLDEzMjAwMCwxMjIwMDAsMjQxMDAwLDY1MDAwXSwib3BzIjoxNjA5MzB9LHsibmFtZSI6Ikl0ZXJhdGlvbi1iYXNlZCIsImNvZGUiOiJkYXRhLmZvckVhY2goc3RyID0%2BIGlzVmFsaWRJbmRleFYzKHN0cikpIiwicnVucyI6WzIwMDAsMzIwMDAsMjIyMDAwLDE5NjAwMCwzMzEwMDAsMzAxMDAwLDEwMDAsMTg4MDAwLDMzMTAwMCwxMTAwMDAsMzA4MDAwLDE1NDAwMCwyNTEwMDAsMTAwMDAwLDUyMDAwLDIwMDAwMCw5OTAwMCw1OTAwMCwxMjAwMCwzMjcwMDAsMjg2MDAwLDM2MzAwMCwyNTEwMDAsMTAwMCwzNjMwMDAsMTY1MDAwLDI3ODAwMCwzMzEwMDAsMjAwMDAwLDUxMDAwLDIzNjAwMCwxMDAwLDEwMDAsMTIzMDAwLDMxNzAwMCw1MzAwMCwzMTQwMDAsMTI1MDAwLDMxNTAwMCwyMjAwMCwzMjcwMDAsMTAwMCwyNzQwMDAsNDkwMDAsMTczMDAwLDI4NTAwMCwxMDAwLDExMjAwMCwyODAwMCwxNDUwMDAsMjQ4MDAwLDE3MzAwMCw1MDAwMCwxMDkwMDAsMzYwMDAwLDI1MzAwMCwxMDAwLDE0MTAwMCwxMDAwLDY1MDAwLDEwMDAsNjYwMDAsMTAwMCwxMDAwLDIwODAwMCwxOTIwMDAsMjQ2MDAwLDMzMTAwMCwyNzkwMDAsMTg1MDAwLDI2NjAwMCwzMzEwMDAsMzMxMDAwLDEzNTAwMCwxMzQwMDAsMTAwMCwyMzgwMDAsNjYwMDAsMTgwMDAsMTAwMCwxMTUwMDAsNzIwMDAsMzMxMDAwLDE0MzAwMCwxODcwMDAsMjM5MDAwLDEwMDAsMTAwMCwxMTAwMDAsMjgzMDAwLDI0MTAwMCwxMzYwMDAsMTUwMDAwLDIyODAwMCwxMTgwMDAsOTAwMDAsNjQwMDAsMzQ1MDAwLDMyMzAwMCwxMDAwXSwib3BzIjoxNjA3NzB9LHsibmFtZSI6InBhcnNlRmxvYXQtYmFzZWQiLCJjb2RlIjoiZGF0YS5mb3JFYWNoKHN0ciA9PiBpc1ZhbGlkSW5kZXhWNChzdHIpKSIsInJ1bnMiOlsyMjMwMDAsMjU4MDAwLDg4MDAwLDE5MDAwLDE3MDAwMCwyNzcwMDAsMjA3MDAwLDE4ODAwMCwzMjMwMDAsMTg0MDAwLDI0MTAwMCwzMDAwLDIyMzAwMCwxNDAwMDAsNzkwMDAsMjE3MDAwLDkzMDAwLDc2MDAwLDIxMDAwMCwxODgwMDAsMjkyMDAwLDE5ODAwMCw0MzAwMCwxMTEwMDAsMjE1MDAwLDEwMDAsMjAyMDAwLDE2NzAwMCwxNzIwMDAsMzUxMDAwLDE2NTAwMCwyNjAwMCwxOTEwMDAsMjI3MDAwLDY0MDAwLDI4MjAwMCwxMDEwMDAsNjMwMDAsMTk2MDAwLDE4ODAwMCwyMTkwMDAsMTAwMCwxNTkwMDAsMTgyMDAwLDExNzAwMCwxNjQwMDAsNDAwMDAsMjM4MDAwLDEyMDAwMCwxOTEwMDAsMjUwMDAsMTE5MDAwLDI4NzAwMCwxODgwMDAsMTU3MDAwLDE1OTAwMCwzMzkwMDAsMTg4MDAwLDUwMDAsMTU5MDAwLDMxMzAwMCwyODQwMDAsMTAwMCw1MjAwMCwxMDAwLDE2NjAwMCwxMjIwMDAsMzQwMDAsMTE3MDAwLDkwMDAsMjMzMDAwLDEyNjAwMCwyMzYwMDAsNTIwMDAsMzIxMDAwLDM1MzAwMCwxMDAwLDY3MDAwLDQ5MDAwLDk4MDAwLDE4NTAwMCwxMjAwMCwyNzAwMDAsMjEwMDAsMTQwMDAwLDE4MzAwMCwyNTAwMDAsMzU2MDAwLDExMDAwMCwxNTAwMDAsMTYyMDAwLDI4NDAwMCw1OTAwMCw2MTAwMCwxODgwMDAsMzE1MDAwLDkyMDAwLDI2OTAwMCwxNjAwMDAsMjUzMDAwXSwib3BzIjoxNTg3NDB9XSwidXBkYXRlZCI6IjIwMjMtMDUtMTFUMTk6NTI6MjMuNjMwWiJ9) the regex-version against other two alternatives. Regex had won. I included your suggestion as a 4th alternative, but regex still wins, so I'm keeping it.

Uzlopak replied:
Usually benchmarking is done in node and not in browser. 

But check [this](https://perf.link/#eyJpZCI6IndjcXoyZDM5OTFwIiwidGl0bGUiOiJDaGVja2luZyBpZiBzdHJpbmcgaXMgYSB2YWxpZCBpbnRlZ2VyIG51bWJlciA%2BPSAwIiwiYmVmb3JlIjoiY29uc3QgZGF0YSA9IFsnMCcsICcwMDEnLCAnMTAwJywgJzEyMzQnLCAnMjInLCAnOTknLCAnYWJjJywgJy0xMjMnLCAnMS43MicsICcxejMnXTtcblxuZnVuY3Rpb24gaXNWYWxpZEluZGV4VjEoc3RyKSB7XG4gIGlmIChzdHIubGVuZ3RoID4gMSAmJiBzdHJbMF0gPT09ICcwJykgcmV0dXJuIGZhbHNlO1xuICByZXR1cm4gL15cXGQrJC8udGVzdChzdHIpO1xufVxuXG5jb25zdCBtdWx0aURpZ2l0TnVtYmVyUkUgPSAvXlxcZCR8XlteMF1cXGQrJC9cbmNvbnN0IGlzVmFsaWRJbmRleFY1ID0gbXVsdGlEaWdpdE51bWJlclJFLnRlc3QuYmluZChtdWx0aURpZ2l0TnVtYmVyUkUpXG4iLCJ0ZXN0cyI6W3sibmFtZSI6IlRlc3QgQ2FzZSIsImNvZGUiOiJkYXRhLmZvckVhY2goc3RyID0%2BIGlzVmFsaWRJbmRleFYxKHN0cikpIiwicnVucyI6WzExMzAwMCwzMTEwMDAsMTAwMCwyMDAwLDEwMDAsNTQwMDAsNjUwMDAsNzIwMDAsODEwMDAsMjcwMDAwLDE1MTAwMCwzMTcwMDAsMTU4MDAwLDEwMDAsMzE3MDAwLDE1NjAwMCwyOTAwMCw5OTAwMCwxMDAwLDIzMDAwLDU5MDAwLDMyMDAwMCwxMDAwLDMwMTAwMCw3OTAwMCwzOTAwMCw1MTAwMCwxMDAwLDU5MDAwLDEwMDAsMTc1MDAwLDYyMDAwLDEwNjAwMCw4MzAwMCwxMDAwLDI2MzAwMCwzMTAwMCwxMDAwLDIyMDAwLDE1MDAwLDE4MDAwMCwxNTgwMDAsMzEwMDAsMTMzMDAwLDExMzAwMCwxMDAwLDcwMDAwLDE4NjAwMCwxNDAwMDAsMTYwMDAsNTYwMDAsOTUwMDAsNjcwMDAsMjkwMDAwLDExNTAwMCwxMDAwLDI1NzAwMCwyMjYwMDAsMTM2MDAwLDkxMDAwLDk2MDAwLDYwMDAwLDEwMDAsMTI0MDAwLDEwMDAsNDAwMCwxMDYwMDAsMjUzMDAwLDEwOTAwMCwzNjAwMCwzMTgwMDAsMjg0MDAwLDE4ODAwMCwxMDAwLDEwMDAsMTU1MDAwLDEwMDAsOTcwMDAsMTQ5MDAwLDIzNDAwMCw1MzAwMCwxNjQwMDAsMTQ5MDAwLDIyMDAwMCwxMDAwLDE3MTAwMCwxMDAwLDQwMDAwLDEyNzAwMCw4NjAwMCwzNDMwMDAsNDkwMDAsMTc1MDAwLDE3MTAwMCw3NzAwMCwyNTMwMDAsMjE0MDAwLDEyMjAwMCwxMzAwMDAsMzIwMDBdLCJvcHMiOjExMDUwMH0seyJuYW1lIjoiVGVzdCBDYXNlIiwiY29kZSI6ImRhdGEuZm9yRWFjaChzdHIgPT4gaXNWYWxpZEluZGV4VjUoc3RyKSkiLCJydW5zIjpbMzE0MDAwLDU3MDAwLDYwMDAsMTk4MDAwLDMyMzAwMCwyODMwMDAsMTQ1MDAwLDM4OTAwMCw0NzAwMCwyNDcwMDAsMjA1MDAwLDQwMDAwLDE4NzAwMCwxMjUwMDAsMjM5MDAwLDUwNTAwMCw0NDAwMCwyMjgwMDAsNDc0MDAwLDEwNDAwMCw5NTAwMCw0NjcwMDAsNTQwMDAsMTA0MDAwLDM1MzAwMCw1NjUwMDAsMjY1MDAwLDEwMDAsMjA5MDAwLDE5ODAwMCwxMDAwLDE5OTAwMCwxNzgwMDAsNTUwMDAsMjI5MDAwLDU3MDAwLDE0NjAwMCwyMzQwMDAsMjkyMDAwLDkyMDAwLDE5MDAwLDExMzAwMCw0MTAwMCwxMzkwMDAsMzg5MDAwLDM1NTAwMCw0MjAwMCw1MjAwMCw5MDAwLDI0MDAwMCwxMTMwMDAsMjkwMDAsMjQ4MDAwLDEwMDAsMjc3MDAwLDIwMDAsMTAwMCwwLDI5MDAwMCwxMDAwLDMwNzAwMCwxMDAwLDEwMDAsMzMwMDAsMzA2MDAwLDQ2NDAwMCw0MTUwMDAsMTg5MDAwLDExOTAwMCwxMDAwLDE0NzAwMCwxNjcwMDAsNTE4MDAwLDQwMTAwMCwxMjIwMDAsMzA4MDAwLDIwMTAwMCwxMDAwLDgwMDAwLDM5MDAwLDI5OTAwMCwxMDAwLDE2NzAwMCwyNTIwMDAsMzY2MDAwLDI0NjAwMCwyMjgwMDAsODQwMDAsNDMwMDAsMTg0MDAwLDEwMDAsMjE2MDAwLDQ3MDAwLDE3MDAwLDE3NjAwMCwxMDAwLDEwNzAwMCwxODAwMCwxMzQwMDAsNDM1MDAwXSwib3BzIjoxNzE1NzB9XSwidXBkYXRlZCI6IjIwMjMtMDUtMTFUMjM6MTM6MzEuNzE4WiJ9) out:

alcidesqueiroz replied:
Your version allows any character as the first one, except for 0. For example, `#123`, `)2222` and `*1` are all valid.

Uzlopak replied:
how about ^\d$|^[1-9]\d+$

alcidesqueiroz replied:
This one works as expected and performs slightly better than the previous one. I've changed it. 👍🏼

Uzlopak commented on index.js:
```diff
@@ -83,6 +94,116 @@ function attachToBody (options, req, reply, next) {
+  const segments = [{ parentRef: body, key: tokens[0] }]
+  const parseArrays = options.enableSpecialNotationKeys !== 'objectsOnly'
+
+  for (let i = 1; i < tokens.length; i++) {
```
Why skipping the first element?

alcidesqueiroz replied:
Because the first token cannot be a '[', ']' or '{}', therefore, it is validated separately, a few lines above this loop. From the second token on, the checks are the same.

Uzlopak commented on index.js:
```diff
@@ -83,6 +94,116 @@ function attachToBody (options, req, reply, next) {
+      value = secureJSON.parse(value)
+    }
+
+    if (parentRef) segments.push({ parentRef, key })
```
```suggestion
    parentRef && segments.push({ parentRef, key })
```

alcidesqueiroz replied:
I've nothing against using short circuit to replace inline ifs, but in this case, I don't see the real benefit. I even [benchmarked it](https://perf.link/#eyJpZCI6IndjcXoyZDM5OTFwIiwidGl0bGUiOiJJZiB2cyBzaG9ydC1jaXJjdWl0IiwiYmVmb3JlIjoibGV0IGZvbyA9IHRydWU7IiwidGVzdHMiOlt7Im5hbWUiOiJJZiIsImNvZGUiOiJpZiAoZm9vKSBjb25zb2xlLmxvZygnYmFyIScpOyIsInJ1bnMiOls3MDAwMCw0ODAwMCw5NjAwMCw3MTAwMCw3NDAwMCw3MTAwMCwxMDIwMDAsMTQ1MDAwLDUwMDAsMTc0MDAwLDExMDAwMCwxMjEwMDAsODgwMDAsMjUwMDAsMTI3MDAwLDE0NTAwMCwxMjMwMDAsMTczMDAwLDgwMDAwLDE4NDAwMCwxNDkwMDAsNjgwMDAsNTEwMDAsMTYwMDAsMTIxMDAwLDEwMzAwMCwxMjEwMDAsNjEwMDAsMTYzMDAwLDQ2MDAwLDE4NTAwMCw1OTAwMCwyMjAwMCw1MDAwLDEzNjAwMCw3NDAwMCw4MjAwMCwxMDAwLDE2MDAwLDc4MDAwLDM1MDAwLDk2MDAwLDE1MDAwLDQ5MDAwLDEwMDAsMTAwMCw2MTAwMCwyMTcwMDAsMTE0MDAwLDUyMDAwLDE1MDAwMCw2NTAwMCw2ODAwMCwyMDAwLDEwMDAsNTMwMDAsMjAwMCwxMDAwMDAsMTUxMDAwLDQwMDAwLDQzMDAwLDUzMDAwLDEzOTAwMCwxMDUwMDAsOTQwMDAsNzAwMDAsNTkwMDAsMTEwMDAwLDkxMDAwLDQ1MDAwLDMwMDAsMTg2MDAwLDY3MDAwLDEwMDAsNzAwMDAsMTMzMDAwLDMxMDAwLDU3MDAwLDY0MDAwLDg4MDAwLDQwMDAwLDQ2MDAwLDEyNzAwMCwxMjEwMDAsNzcwMDAsOTcwMDAsNzAwMCwzNzAwMCw4MDAwMCwyNDAwMCwxNDEwMDAsNjcwMDAsMjEwMDAwLDEwMDAsNzEwMDAsNzkwMDAsMTIwMDAwLDc1MDAwLDIxOTAwMCwxNzAwMDBdLCJvcHMiOjgxODAwfSx7Im5hbWUiOiJTaG9ydC1jaXJjdWl0IiwiY29kZSI6ImZvbyAmJiBjb25zb2xlLmxvZygnYmFyIScpOyIsInJ1bnMiOlsxODAwMCwxNjQwMDAsMzAwMDAsMzAwMCw3NzAwMCw3MTAwMCwyMjAwMDAsNzgwMDAsMTQ3MDAwLDEwMDAsNTAwMDAsMzUwMDAsODUwMDAsMjUwMDAsMTY3MDAwLDEzODAwMCw2NzAwMCw0NDAwMCw2NjAwMCw2MDAwLDE0MDAwMCwxOTAwMDAsMTgwMDAsNzEwMDAsMTIzMDAwLDQ1MDAwLDExNjAwMCwxNzgwMDAsOTIwMDAsMTY5MDAwLDEwODAwMCwyMzAwMCwzMTAwMCw3NjAwMCw5MDAwLDEwMDAsMTg1MDAwLDEwMDAsNTcwMDAsMTI2MDAwLDcwMDAwLDEwMTAwMCwxMDQwMDAsMTAwMCwxNTgwMDAsOTgwMDAsNTUwMDAsMzEwMDAsNTcwMDAsMzcwMDAsNzEwMDAsMTQ2MDAwLDQwMDAwLDI3MDAwLDU0MDAwLDE1OTAwMCwyNzAwMCwxOTIwMDAsMzMwMDAsNjcwMDAsMTU3MDAwLDEwMDAwLDMwMDAwLDEwMDAsODcwMDAsMTUwMDAsMjYwMDAsNDAwMDAsMTM0MDAwLDEwMDAsMTAzMDAwLDMzMDAwLDUwMDAsMTA3MDAwLDEzMzAwMCwyNTAwMCwxMzcwMDAsMTIwMDAsMTg1MDAwLDE1MjAwMCwyMTAwMDAsMjMwMDAsODYwMDAsMjAwMDAsNTkwMDAsNzgwMDAsMTIwMDAwLDY4MDAwLDEwOTAwMCw0NzAwMCw5NjAwMCwxODQwMDAsMTAwMCwxMDEwMDAsNzkwMDAsNjUwMDAsODgwMDAsNDAwMCwyMTEwMDAsMTIxMDAwXSwib3BzIjo3OTQyMH1dLCJ1cGRhdGVkIjoiMjAyMy0wNS0xMVQyMDoxODoxMC44MzBaIn0%3D), but the results are not conclusive. If I rerun the benchmark multiple times, sometimes `if` wins by a slim margin, other times short circuit is slightly faster. So, I don't see why we should change it.

Uzlopak replied:
Then please add some curly brackets please ;)

Uzlopak commented on index.js:
```diff
@@ -83,6 +94,116 @@ function attachToBody (options, req, reply, next) {
+  // only a single {} special ending is allowed
+  else if (tokens.filter(t => t === '{}').length > 1) isValid = false
+  // {}, if present, must be the last token
+  else if (tokens.find(t => t === '{}') && tokens[tokens.length - 1] !== '{}') isValid = false
```
we checked already in line 108, that the token is only existing once. Maybe it is "cheaper" to store the count once and check if the last token is '{}'

Like this


```js
  let amountOfObjectNotation = 0

  // first token can't be '[', ']', or '{}'
  if (['[', ']', '{}'].includes(tokens[0])) isValid = false
  // only a single {} special ending is allowed
  else if ((amountOfObjectNotation = tokens.filter(t => t === '{}').length) > 1) isValid = false
  // {}, if present, must be the last token
  else if (amountOfObjectNotation === 1 && tokens[tokens.length - 1] !== '{}') isValid = false
```

Not tested, maybe wrong.

Uzlopak commented on index.js:
```diff
@@ -83,6 +94,116 @@ function attachToBody (options, req, reply, next) {
+  let isValid = true
+
+  // first token can't be '[', ']', or '{}'
+  if (['[', ']', '{}'].includes(tokens[0])) isValid = false
```
Maybe we can extract this Array outside the function, as we can save to instantiate it with every function call.

Uzlopak commented on index.js:
```diff
@@ -83,6 +94,116 @@ function attachToBody (options, req, reply, next) {
+
+  let hasOpenBracket = false
+  for (let i = 1; i < tokens.length; i++) {
+    const token = tokens[i]
```
maybe define token and nextToken as let outside the for loop. Avoiding variable instantiation by reusing the variables.


Someting like:
```js
const tokensLength = tokens.length
let token = tokens[0]
let nextToken = tokens[1]
let i = 1
 while (i < tokensLength) {
     token = nextToken
     nextToken = tokes[++i]
```

alcidesqueiroz replied:
Even though, IMO, this approach seems uglier, I have to agree that [it performs better](https://perf.link/#eyJpZCI6IndjcXoyZDM5OTFwIiwidGl0bGUiOiJSZWd1bGFyIGZvciBsb29wICsgYmxvY2sgbGV2ZWwgY29uc3RzIFZTIHdoaWxlIGxvb3AgKyBtYW51YWxseSBob2lzdGluZyB2YXJzIiwiYmVmb3JlIjoiY29uc3QgZGF0YSA9IG5ldyBBcnJheSgxMDAwKS5maWxsKDApLm1hcCgoKSA9PiBNYXRoLnJhbmRvbSgpLnRvU3RyaW5nKCkpOyIsInRlc3RzIjpbeyJuYW1lIjoiUmVndWxhciBmb3IgbG9vcCArIGJsb2NrIGxldmVsIGNvbnN0cyIsImNvZGUiOiJmb3IgKGxldCBpID0gMDsgaSA8IGRhdGEubGVuZ3RoOyBpKyspIHtcbiAgY29uc3QgdmFsdWUgPSBkYXRhWzBdO1xuICBjb25zdCBuZXh0VmFsdWUgPSBkYXRhWzFdO1xufSIsInJ1bnMiOls2MDAwLDIwMDAsNjAwMCwzMDAwLDIwMDAsNTAwMCwxODAwMCwxMjAwMCwxMzAwMCw1MDAwLDYwMDAsMTEwMDAsMTcwMDAsNTAwMCw2MDAwLDIyMDAwLDIzMDAwLDE1MDAwLDYwMDAsMTAwMCw3MDAwLDYwMDAsMTAwMDAsMjQwMDAsMTcwMDAsMjAwMCw3MDAwLDIwMDAwLDEwMDAsMTQwMDAsNTAwMCw5MDAwLDYwMDAsNTAwMCwxNzAwMCwxMjAwMCwxODAwMCwxNjAwMCwxMzAwMCw3MDAwLDUwMDAsNTAwMCwxMzAwMCw0MDAwLDYwMDAsMTkwMDAsMzAwMCwxMjAwMCwxODAwMCwxMTAwMCw3MDAwLDIzMDAwLDQwMDAsMTAwMCw5MDAwLDEyMDAwLDE2MDAwLDIzMDAwLDE3MDAwLDEwMDAsMTAwMCwxNjAwMCw2MDAwLDIwMDAwLDEwMDAwLDkwMDAsMTMwMDAsMTYwMDAsMTAwMDAsMTAwMDAsNjAwMCw2MDAwLDE0MDAwLDQwMDAsNzAwMCw3MDAwLDkwMDAsNTAwMCw2MDAwLDE2MDAwLDQwMDAsNjAwMCw0MDAwLDIwMDAsMTAwMDAsOTAwMCw4MDAwLDUwMDAsNjAwMCwyMjAwMCwxMDAwLDExMDAwLDI0MDAwLDIxMDAwLDE3MDAwLDIwMDAsNDAwMCw1MDAwLDIxMDAwLDQwMDBdLCJvcHMiOjk4NjB9LHsibmFtZSI6IndoaWxlIGxvb3AgKyBtYW51YWxseSBob2lzdGluZyB2YXJzIiwiY29kZSI6ImNvbnN0IGRhdGFMZW5ndGggPSBkYXRhLmxlbmd0aDtcbmxldCB2YWx1ZSA9IGRhdGFbMF07XG5sZXQgbmV4dFZhbHVlID0gZGF0YVsxXTtcbmxldCBpID0gMTtcbndoaWxlIChpIDwgZGF0YUxlbmd0aCkge1xuICAgdmFsdWUgPSBkYXRhW2ldO1xuICAgbmV4dFZhbHVlID0gZGF0YVsrK2ldO1xufSIsInJ1bnMiOls0MDAwLDkwMDAsODAwMCw3MDAwLDkwMDAsNDAwMCw5MDAwLDIxMDAwLDQwMDAsMTIwMDAsMTkwMDAsMTAwMDAsOTAwMCwyMTAwMCw4MDAwLDgwMDAsNTAwMCwxODAwMCwyNjAwMCwyNjAwMCwxNzAwMCwxMDAwMCwxNTAwMCwyMzAwMCw0MDAwLDE3MDAwLDIwMDAsMjAwMDAsMTMwMDAsMTcwMDAsOTAwMCwyMTAwMCwyMDAwLDI1MDAwLDEzMDAwLDcwMDAsMjAwMDAsOTAwMCwxNjAwMCw5MDAwLDQwMDAsNjAwMCwyNDAwMCw3MDAwLDI1MDAwLDEwMDAwLDgwMDAsMTUwMDAsNDAwMCw5MDAwLDYwMDAsMTAwMCwxNTAwMCwxMzAwMCwxMjAwMCw5MDAwLDMwMDAsMjUwMDAsMTgwMDAsODAwMCwxMDAwLDYwMDAsOTAwMCwyMDAwLDEwMDAsMTkwMDAsNDAwMCwxNjAwMCwzMDAwLDEyMDAwLDI1MDAwLDkwMDAsMjEwMDAsNDAwMCwyNDAwMCwzMDAwLDE4MDAwLDQwMDAsMjIwMDAsMTQwMDAsMzAwMCw5MDAwLDkwMDAsMTUwMDAsMzAwMCw4MDAwLDkwMDAsMTgwMDAsNjAwMCwyMzAwMCw5MDAwLDYwMDAsMjkwMDAsNzAwMCwxMDAwLDE0MDAwLDEyMDAwLDIxMDAwLDExMDAwLDQwMDBdLCJvcHMiOjExNjIwfV0sInVwZGF0ZWQiOiIyMDIzLTA1LTExVDIyOjMzOjE2LjM2MFoifQ%3D%3D). And since performance is a main concern for Fastify and its plugins, I've changed it.

Uzlopak reviewed (commented):
Had some thouhts on the code.

Uzlopak reviewed (commented):
Had some thouhts on the code.

alcidesqueiroz commented on index.js:
```diff
@@ -73,7 +73,18 @@ function attachToBody (options, req, reply, next) {
     }
-    if (body[key] === undefined) {
+
+    const tokens = [true, 'objectsOnly'].includes(options.enableSpecialNotationKeys) ? getSerializedKeyTokens(key) : []
```
Good suggestion. I usually prefer `includes`, as it makes expressions more concise, but performance-wise, chaining ORs definitely wins ([benchmarked both](https://perf.link/#eyJpZCI6IndxOTNoM3puYjJkIiwidGl0bGUiOiJpZiBzdGF0ZW1lbnQgdXNpbmcgYXJyYXkuaW5jbHVkZXMgdnMgT1IiLCJiZWZvcmUiOiJjb25zdCBkYXRhID0gW2ZhbHNlLCAxLCAnZm9vJywgJ2JhcicsICd4JywgMTIzLCBudWxsLCAnb2JqZWN0c09ubHknLCB1bmRlZmluZWQsIHRydWUsICd6enp6eiddIiwidGVzdHMiOlt7Im5hbWUiOiJVc2luZyBpbmNsdWRlcyIsImNvZGUiOiJkYXRhLmZvckVhY2goKHZhbHVlKSA9PiB7IFxuICBpZiAoW3RydWUsICdvYmplY3RzT25seScsIDEyMywgJ3l5eSddLmluY2x1ZGVzKHZhbHVlKSkge1xuICAgIGNvbnNvbGUubG9nKCdvayEnKTtcbiAgfVxufSkiLCJydW5zIjpbMjIwMDAsMzEwMDAsNjAwMDAsMzEwMDAsMjAwMDAsMTgwMDAsMTUwMDAsMjYwMDAsNDAwMCwxMzAwMCw3MDAwLDEwMDAsMzcwMDAsMzEwMDAsNTIwMDAsMzQwMDAsMTgwMDAsMjEwMDAsNTUwMDAsMjcwMDAsMTQwMDAsMTgwMDAsMTcwMDAsMTAwMCw0MTAwMCw0NTAwMCwyNjAwMCw0NDAwMCw1MDAwLDI4MDAwLDMwMDAwLDE5MDAwLDExMDAwLDIwMDAsNDAwMCw0NzAwMCw0MDAwMCwyNTAwMCw2NjAwMCwzODAwMCw2MjAwMCw0MDAwMCwxMTAwMCwyMDAwLDE5MDAwLDIxMDAwLDYzMDAwLDIyMDAwLDEwMDAsMTAwMCw1MDAwMCw0MzAwMCwzMzAwMCw5MDAwLDgwMDAsNTAwMCw0MjAwMCw5MDAwLDUwMDAsMjQwMDAsNzAwMCw4MDAwLDU0MDAwLDEyMDAwLDQ1MDAwLDI0MDAwLDQxMDAwLDU0MDAwLDQwMDAwLDI2MDAwLDI3MDAwLDMzMDAwLDE0MDAwLDEwMDAsMjMwMDAsNjAwMCwxOTAwMCw0MzAwMCwxMTAwMCw2MDAwLDIyMDAwLDIyMDAwLDQwMDAsMjAwMCwxOTAwMCwxNzAwMCwyNjAwMCwyNDAwMCw2MDAwLDQ3MDAwLDM2MDAwLDQwMDAsMTkwMDAsODAwMCwyMzAwMCwxMDAwLDIzMDAwLDEwMDAsNTcwMDAsMjAwMDBdLCJvcHMiOjIzODkwfSx7Im5hbWUiOiJVc2luZyBPUiIsImNvZGUiOiJkYXRhLmZvckVhY2goKHZhbHVlKSA9PiB7IFxuICBpZiAodmFsdWUgPT09IHRydWUgfHwgdmFsdWUgPT09ICdvYmplY3RzT25seScgfHwgdmFsdWUgPT09IDEyMyB8fCB2YWx1ZSA9PT0gJ3l5eScpIHtcbiAgICBjb25zb2xlLmxvZygnb2shJyk7XG4gIH1cbn0pIiwicnVucyI6WzU4MDAwLDI3MDAwLDUxMDAwLDE3MDAwLDQyMDAwLDE3MDAwLDM5MDAwLDQwMDAsMzQwMDAsMjAwMCw2MTAwMCwxMDAwMCw0ODAwMCwzMTAwMCwxNjAwMCw0MTAwMCwxNzAwMCw0MjAwMCwxODAwMCw0MTAwMCwzNTAwMCw3MzAwMCwxNTAwMCwxNzAwMCwzNTAwMCwxMjAwMCwzNTAwMCwxMDAwLDQxMDAwLDgwMDAsNjIwMDAsMjUwMDAsMTQwMDAsMzIwMDAsNDAwMDAsMzUwMDAsMjUwMDAsOTAwMCwzNDAwMCw3MjAwMCwyMDAwLDQ1MDAwLDEwMDAsMTAwMCwyMDAwMCwxMjAwMCwxNjAwMCw0NDAwMCw2MDAwLDMzMDAwLDIxMDAwLDQxMDAwLDM1MDAwLDU2MDAwLDYyMDAwLDM2MDAwLDI5MDAwLDI3MDAwLDM0MDAwLDE0MDAwLDMwMDAwLDUxMDAwLDkwMDAsMzYwMDAsMjcwMDAsNjUwMDAsMTEwMDAsNjQwMDAsNTYwMDAsNTUwMDAsNTIwMDAsNDAwMCwxMDAwLDkwMDAsMzYwMDAsMzEwMDAsMTkwMDAsMjEwMDAsNzIwMDAsNjEwMDAsMTYwMDAsMzkwMDAsODEwMDAsMTQwMDAsNDcwMDAsNTgwMDAsMzAwMDAsODAwMCwzNzAwMCw3MDAwLDMzMDAwLDUxMDAwLDEzMDAwLDQyMDAwLDgxMDAwLDM0MDAwLDYwMDAsODAwMCw2MDAwLDYyMDAwXSwib3BzIjozMTU0MH1dLCJ1cGRhdGVkIjoiMjAyMy0wNS0xMVQxNzo0NjoyNS42MDRaIn0%3D), and the OR version was ~33% faster).

Uzlopak commented on index.js:
```diff
@@ -83,6 +99,133 @@ function attachToBody (options, req, reply, next) {
   })
 }
 
+const allowedTokens = ['[', ']', '{}']
```
actually those are "notAllowedAsFirstTokens"

alcidesqueiroz replied:
Actually, since this variable was being used in two places, this name wouldn't make sense in both contexts. If I was going to keep it, I would rename it to `specialTokens`, but I decided to simply remove it, as I've replaced the calls to `.includes` with chained ORs.

Uzlopak commented on index.js:
```diff
@@ -73,7 +73,23 @@ function attachToBody (options, req, reply, next) {
     }
-    if (body[key] === undefined) {
+
+    const tokens = (
```
attachToBody is called only in the preValidation hook in line 277. We could actually add above the addHook a line to normalize the value. Then we dont need to check here for multiple values.

alcidesqueiroz replied:
Didn't get it. Could you explain it in more details?

Uzlopak commented on index.js:
```diff
@@ -83,6 +99,133 @@ function attachToBody (options, req, reply, next) {
+const allowedTokens = ['[', ']', '{}']
+
+function getSerializedKeyTokens (key) {
+  const tokens = key.split(/(\[|]|{})/).filter(Boolean)
```
We split. But why do we filter?

alcidesqueiroz replied:
This is for filtering out empty strings. It could've been written as `key.split(/(\[|]|{})/).filter(t => t)` too.

Uzlopak commented on index.js:
```diff
@@ -83,6 +99,133 @@ function attachToBody (options, req, reply, next) {
+
+function getSerializedKeyTokens (key) {
+  const tokens = key.split(/(\[|]|{})/).filter(Boolean)
+  return tokens.some(token => allowedTokens.includes(token)) ? tokens : []
```
.some() is potentially slower than a for loop.

alcidesqueiroz replied:
Replaced it with a negative while, which was the fastest alternative.

Uzlopak commented on index.js:
```diff
@@ -83,6 +99,133 @@ function attachToBody (options, req, reply, next) {
+  else if (token === '__proto__' || token === 'constructor') {
+    return `${token} is not allowed as a key name`
+  } else {
+    const bracesEndingCount = tokens.filter(t => t === '{}').length
```
this is definetly slower than a for loop

alcidesqueiroz replied:
I haven't benchmarked it on Node, but [this is what I got on Chrome 112.0.5615.137](https://perf.link/#eyJpZCI6InN1NzRlZ3Q1NjJsIiwidGl0bGUiOiJHZXR0aW5nIGl0ZW0gY291bnQ6IGZpbHRlciBWUyBmb3IgVlMgd2hpbGUiLCJiZWZvcmUiOiJjb25zdCBkYXRhID0gWydwcm9wQVtwcm9wQUFdJywgJ3Byb3BBW3Byb3BBQl0nLCAncHJvcEFbcHJvcEFDXVtdJywgJ3Byb3BBW3Byb3BBQ11bXScsICdwcm9wQVtwcm9wQUNdW10nLCAncHJvcEJbMF1bcHJvcEJBXVtwcm9wQkFBXScsICdwcm9wQlswXVtwcm9wQkFdW3Byb3BCQUJdJywgJ3Byb3BCWzBdW3Byb3BCQV1bcHJvcEJBQ11bMF0nLCAncHJvcEJbMF1bcHJvcEJBXVtwcm9wQkFDXVsxXScsICdmb297fVthYWFde31bYmFyXScsICdwcm9wQlswXVtwcm9wQkFdW3Byb3BCQUNdWzJdJywgJ3Byb3BCWzBdW3Byb3BCQl0nLCAncHJvcEJbMV1bcHJvcEJBXVtwcm9wQkFBXScsICdwcm9wQ1tdJywgJ3Byb3BDW10nLCAncHJvcENbXScsICdwcm9wRCcsICdwcm9wRScsICdwcm9wRScsICdwcm9wRScsICdwcm9wRnt9JywgJ3Byb3BHWzBdW3Byb3BHQV1bXXt9JywgJ3Byb3BHWzFde30nLCAnYWFhe317fSddOyIsInRlc3RzIjpbeyJuYW1lIjoiZm9yIiwiY29kZSI6ImxldCBicmFjZXNFbmRpbmdDb3VudCA9IDBcbmxldCB0b2tlbjtcblxuZm9yIChsZXQgaSA9IDA7IGkgPCBkYXRhLmxlbmd0aDsgaSsrKSB7XG4gIHRva2VuID0gZGF0YVtpXVxuICBpZiAodG9rZW4gPT09ICd7fScpIHtcbiAgICBicmFjZXNFbmRpbmdDb3VudCsrXG4gIH1cbn0iLCJydW5zIjpbMjMzMDAwLDI4MjAwMCwxOTQwMDAsMzAwMDAwLDMyOTAwMCwyMDgwMDAsMzQ4MDAwLDE0MzAwMCwyODMwMDAsMCwxNjkwMDAsMzYzMDAwLDgwMDAsMzAwMDAwLDcxNjAwMCwyODAwMCwxNzEwMDAsNDYwMDAsNTUyMDAwLDI4NjAwMCwyNzUwMDAsODMxMDAwLDIwNzAwMCw1MTIwMDAsNTEzMDAwLDMxNzAwMCwxODEwMDAsMjMwMDAwLDYxMDAwLDExNzAwMCwzMzcwMDAsNTc5MDAwLDE5NjAwMCwzMTQwMDAsMzQ5MDAwLDI4MjAwMCw2NDUwMDAsMTkxMDAwLDE4MTAwMCwyNjIwMDAsMzAwMDAsMzcwMDAsMzAwMDAwLDUwOTAwMCw3MjAwMDAsMzAwMDAwLDEwMDAsNDQ0MDAwLDQwNzAwMCw2MTAwMCwzMDAwMDAsNjM3MDAwLDUwMDAwLDYwMDAwMCwzMDAwLDI2OTAwMCw2NDEwMDAsMzUwMDAwLDQwMTAwMCwzMTAwMCw1NjAwMCw1MjAwMCwyOTEwMDAsNDI5MDAwLDEwOTAwMCwzMDAwMDAsMjY0MDAwLDQxODAwMCwxMTEwMDAsMzAwMDAwLDU5MDAwLDM4OTAwMCwyNzcwMDAsNzIwMDAsMzAwMDAwLDc1MzAwMCwzMDAwMDAsNDg1MDAwLDI4NTAwMCwyNzMwMDAsMjQxMDAwLDE5MDAwLDU3MjAwMCwzMDAwMDAsMzAwMDAwLDc5MDAwLDQxMjAwMCw3NjAwMCwzMDAwMDAsMTY4MDAwLDgzMDAwLDgwMDAsNDI4MDAwLDI0MTAwMCw2ODAwMCwyNDAwMDAsMjMzMDAwLDEzMzAwMCwyMDMwMDAsNTQ3MDAwXSwib3BzIjoyNzgwNDB9LHsibmFtZSI6ImZpbHRlciIsImNvZGUiOiJkYXRhLmZpbHRlcih0ID0%2BIHQgPT09ICd7fScpLmxlbmd0aCIsInJ1bnMiOls3MzAwMDAsOTAwMDAsMjkzMDAwLDI3NDAwMCw1MjAwMCw3MTMwMDAsNTE4MDAwLDI1OTAwMCwzODgwMDAsMjMwMDAwLDIzNTAwMCw1NzQwMDAsMjg4MDAwLDY0MDAwMCw4NTYwMDAsMzc4MDAwLDE5NjAwMCw1MjAwMCw3NTkwMDAsMzExMDAwLDMwMDAwMCwzNzkwMDAsODUwMDAwLDY1MzAwMCwxODAwMCw4NDAwMCw5NTAwMCw2ODEwMDAsNjE2MDAwLDcwNDAwMCw3MDAwLDUyMTAwMCwxMjkwMDAsMjY0MDAwLDM2MDAwMCwyOTUwMDAsNzAxMDAwLDU4MDAwLDI4ODAwMCwzODAwMDAsMjk5MDAwLDEzNTAwMCwyMDMwMDAsNjkxMDAwLDczMTAwMCw0MTIwMDAsNDQwMDAsNjA0MDAwLDM0MjAwMCwxOTAwMCw4NDQwMDAsMjY2MDAwLDc0ODAwMCwxODAwMCw2NTcwMDAsNDYxMDAwLDY5MzAwMCwzOTYwMDAsMTcxMDAwLDYxMjAwMCw2MTAwMCwyMTAwMCwzMzIwMDAsNzY3MDAwLDIyNzAwMCw0MDYwMDAsNDc4MDAwLDU4MzAwMCw3MDAwMCwzOTQwMDAsNDYwMDAsODk1MDAwLDE0OTAwMCw2NzAwMCwyNTMwMDAsMjUyMDAwLDQ3ODAwMCwxMjAwMDAsNTIwMDAsNjE0MDAwLDg1MjAwMCwyMTAwMCw3MDAwMDAsMzkxMDAwLDc4NTAwMCwxMTAwMCw3NzMwMDAsODczMDAwLDQ3MDAwMCw3ODQwMDAsMzAwMCwxMzAwMCw0NzcwMDAsMTkwMDAsNTQ4MDAwLDUzODAwMCw2NDYwMDAsMTY1MDAwLDMwNDAwMCwxMDQwMDBdLCJvcHMiOjM4MzA3MH0seyJuYW1lIjoibmVnYXRpdmUgd2hpbGUiLCJjb2RlIjoibGV0IGkgPSBkYXRhLmxlbmd0aCAtIDFcbmxldCB0b2tlbiA9IGRhdGFbaV1cbmxldCBicmFjZXNFbmRpbmdDb3VudCA9IDBcblxud2hpbGUgKGkgPj0gMCkge1xuICBpZiAodG9rZW4gPT09ICd7fScpIHtcbiAgICBicmFjZXNFbmRpbmdDb3VudCsrXG4gIH1cbiAgdG9rZW4gPSBkYXRhWy0taV1cbn0iLCJydW5zIjpbOTYwMDAsNjEwMDAwLDI3MDAwMCw2MDEwMDAsNDAwMDAsNDI0MDAwLDM0NDAwMCw4NDAwMCwyNDUwMDAsNDU3MDAwLDE1ODAwMCwyODQwMDAsMzk2MDAwLDMyNTAwMCw1NjUwMDAsMzk2MDAwLDE0NjAwMCwzNTIwMDAsMzA3MDAwLDU0MDAwMCw0ODkwMDAsMzY1MDAwLDM3MDAwLDI3MDAwMCwyNjIwMDAsMzk2MDAwLDY2MDAwLDM5NjAwMCwzMjIwMDAsODAwMCwxNTUwMDAsMTUwMDAsMTQ3MDAwLDM3NjAwMCwyMDQwMDAsNTkwMDAsMjAwMCw1NDkwMDAsMTA2MDAwLDk0MDAwLDM1MjAwMCw4NzAwMCwxNDAwMDAsNDQyMDAwLDU1NzAwMCwzOTgwMDAsMjMwMDAsMjIwMDAwLDE3MjAwMCw1NzAwMDAsMTczMDAwLDEwMTAwMCwyMjUwMDAsMzc2MDAwLDI1NjAwMCwzMTkwMDAsMjUxMDAwLDgxMDAwLDUwNTAwMCwyMzMwMDAsMzE3MDAwLDE5MzAwMCwzOTYwMDAsNDk5MDAwLDQyMDAwLDIzMTAwMCwyODAwMDAsMzk2MDAwLDYwMjAwMCwzOTkwMDAsMzc1MDAwLDM5NjAwMCw2NDIwMDAsNjY2MDAwLDM3MzAwMCwyNzUwMDAsNDEwMDAwLDMyOTAwMCwyODEwMDAsMjAwMCw1MzkwMDAsNTY1MDAwLDQxNjAwMCwzMTcwMDAsMzk2MDAwLDM5NjAwMCwzNzQwMDAsMzEyMDAwLDY0MjAwMCw0NTkwMDAsMjgwMDAsNTA1MDAwLDM5NjAwMCw1MzAwMCwzMzAwMCw3OTAwMCwyNTAwMDAsMzYwMDAsMTkwMDAwLDk1MDAwXSwib3BzIjoyOTYyNDB9XSwidXBkYXRlZCI6IjIwMjMtMDUtMTJUMTk6MDg6NDYuMTQ4WiJ9). I've replaced it with a negative `while` anyway, as I noticed that array methods like `filter`, `find`, `some` or `every` are not used in the project.

Uzlopak commented on index.js:
```diff
@@ -83,6 +99,133 @@ function attachToBody (options, req, reply, next) {
+  let token = tokens[0]
+
+  // first token can't be '[', ']', or '{}'
+  if (allowedTokens.includes(token)) isValid = false
```
can you use curly brackets please?

Uzlopak commented on index.js:
```diff
@@ -83,6 +99,133 @@ function attachToBody (options, req, reply, next) {
+    else if (bracesEndingCount > 0 && tokens[tokens.length - 1] !== '{}') isValid = false
+  }
+
+  const tokensLen = tokens.length
```
can we skip this whole part, if isValid is set to false?

alcidesqueiroz replied:
I removed the whole `isValid` approach. Now, I'm promptly returning when a key is invalid.

Uzlopak commented on index.js:
```diff
@@ -83,6 +99,133 @@ function attachToBody (options, req, reply, next) {
+  // when the {} special ending is supplied, the value must be valid JSON
+  if (isValid && tokens[tokens.length - 1] === '{}') {
+    try {
+      secureJSON.parse(value)
```
Your examples are `form.append('foo{}', '{"bar":[1,2,3]}') `and more complex keys. But what if the value is a primitive, like 'a' or worse 'null'?

secureParse would not fail in case of null.

alcidesqueiroz replied:
I've included an explicit check for `null`, but `secureJSON.parse` would, in fact, allow primitives to be passed. Is there a good reason for explicitly preventing this?

Uzlopak commented on index.js:
```diff
@@ -83,6 +99,133 @@ function attachToBody (options, req, reply, next) {
+
+function getSerializedKeySegments (options, body, tokens, value) {
+  const segments = [{ parentRef: body, key: tokens[0] }]
+  const parseArrays = options.enableSpecialNotationKeys !== 'objectsOnly'
```
what if enableSpecialNotationKeys is true? Maybe we should normalize like described in a different comment.

alcidesqueiroz replied:
At this point, `enableSpecialNotationKeys` can only be `true` or `"objectsOnly"`, as we're explicitly checking it on line 77. So, if it isn't `"objectsOnly"`, it necessarily must be `true`.

Uzlopak replied:
options.enableSpecialNotationKeys !== 'objectOnly' would be true if enableSpecialNotationKeys is true. So parseArrays would be true. So you would generate Arrays and not Objects with numeric keys?!

alcidesqueiroz replied:
Exactly. When `enableSpecialNotationKeys` is `true`, numeric keys generate arrays. When `enableSpecialNotationKeys` is `objectsOnly`, everything is parsed as an object, so no arrays at all (unless you use the `{}` special ending and declare an array using JSON).

Uzlopak commented on index.js:
```diff
@@ -83,6 +99,133 @@ function attachToBody (options, req, reply, next) {
+    if (token === ']' && prevToken === '[') {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? (parseArrays ? [] : {})
+      key = parseArrays ? parentRef.length : Object.keys(parentRef).length
+    } else if (parseArrays && token === ']' && isValidIndex(prevToken)) {
```
maybe write two while loops, one for the case parseArrays is true and one for the case, that parseArrays is false?

alcidesqueiroz replied:
I'm really uncertain whether this kind of change would be worth it. After all, we would sacrifice simplicity, readability and maintainability to favor small performance gains.

Uzlopak commented on index.js:
```diff
@@ -73,7 +73,23 @@ function attachToBody (options, req, reply, next) {
+        return
+      }
+      const segments = getSerializedKeySegments(options, body, tokens, value)
+      setSerializedKey(segments)
```
So we create segments and then pass it directly to setSerializedKey. It feels somehow that there is an artificial separation.

alcidesqueiroz replied:
Even though this algorithm could be simplified, this format separates concerns in distinct steps (tokenization, validation, segment extraction and appending keys to the body).

alcidesqueiroz commented:
@Uzlopak @climba03003 hey there! 👋🏼 I wanted to check in on the status of this PR. Could you please let me know if there's anything else I need to address or if there's any further feedback? Thanks.

simoneb commented:
@alcidesqueiroz thanks for your contribution here, and apologies for the lack of feedback. I think this was a valuable contribution and it's a pity that we couldn't integrate it in the codebase.

alcidesqueiroz commented:
@simoneb no worries, mate! 🙂 🙏
